src/mjan.c

```diff
@@ -116,16 +120,45 @@ MOVELIST *GenerateMoves(POSITION position) {
   generic_hash_unhash(position, board);
 
   char piece = (generic_hash_turn(position) == 1) ? WCHAR : BCHAR;
-  int diffs[4] = {4, -4, 1, -1}; // up down right left
+  int diffs[8] = {4, -4, 1, -1, 5, -5, 3, -3};
+
+  // Main diagonal positions
+  BOOLEAN isMainDiagonalPos[16] = {
+    TRUE, FALSE, FALSE, TRUE,   // 1, 4
+    FALSE, TRUE, TRUE, FALSE,   // 6, 7
+    FALSE, TRUE, TRUE, FALSE,   // 10, 11
+    TRUE, FALSE, FALSE, TRUE    // 13, 16
+  };
 
-  for (int idx = 0; idx < 16; idx ++) {
+  for (int idx = 0; idx < 16; idx++) {
     if (board[idx] != piece)
       continue;
 
-    for (int cnt = 0; cnt < 4; cnt ++) {
-      int jdx = idx + diffs[cnt];
-      if (jdx >= 0 && jdx < BOARDSIZE && board[jdx] == ECHAR && ((diffs[cnt] == -4 || diffs[cnt] == 4) || jdx / 4 == idx / 4))
-        moves = CreateMovelistNode(ENCODEMOVE(idx, diffs[cnt]), moves);
+    for (int cnt = 0; cnt < 8; cnt++) {
+      int diff = diffs[cnt];
+      int jdx = idx + diff;
+
+      if (jdx < 0 || jdx >= BOARDSIZE)
+        continue;
+
+      if ((diff == 1 || diff == -1) && jdx / 4 != idx / 4)
+        continue;
+
+      if ((diff == 5 || diff == -5 || diff == 3 || diff == -3) &&
+          abs(idx % 4 - jdx % 4) > 1)
+        continue;
+
+      if ((diff == 5 || diff == -5 || diff == 3 || diff == -3)) {
```

Comment:
more efficient to move before if with `abs` call, also please refactor into one statement for readability

Reply:
done.